topdown/subset.go

```diff
@@ -0,0 +1,206 @@
+		// them normally. If only one term is an object, then we
+		// do a normal comparison which will come up false.
+		if ok, superObj, subObj := bothObjects(ast.NewTerm(superVal.Value), ast.NewTerm(subVal.Value)); ok {
+			return objectSubset(superObj, subObj)
+		}
+
+		if ok, superSet, subSet := bothSets(ast.NewTerm(superVal.Value), ast.NewTerm(subVal.Value)); ok {
+			return setSubset(superSet, subSet)
+		}
+
+		if ok, superArray, subArray := bothArrays(ast.NewTerm(superVal.Value), ast.NewTerm(subVal.Value)); ok {
+			return arraySubset(superArray, subArray)
+		}
+
+		if !subVal.Equal(superVal) {
+			return false
+		}
+	}
+
+	return true
+}
+
+// setSubset implements the subset operation on sets.
+//
+// Unlike in the object case, this is not recursive, we just dumbly compare
+// values using ast.Set.Contains() because we have no well defined way to
+// "match up" objects that are in different sets.
+func setSubset(super ast.Set, sub ast.Set) bool {
+	isSubset := true
+	sub.Until(func(t *ast.Term) bool {
+		if !super.Contains(t) {
+			isSubset = false
+			return true
+		}
+		return false
+	})
+
+	return isSubset
+}
+
+// arraySubset implements the subset operation on arrays.
+//
+// This is defined to mean that the entire "sub" array must appear in
+// the "super" array. For the same rationale as setSubset(), we do not attempt
+// to recurse into values.
+func arraySubset(super, sub *ast.Array) bool {
+	// Notice that this is essentially string search. The naive approach
+	// used here is O(n^2). This should probably be rewritten later to use
+	// Boyer-Moore or something.
+
+	if sub.Len() > super.Len() {
+		return false
+	}
+
+	if sub.Equal(super) {
+		return true
+	}
+
+	superCursor := 0
+	subCursor := 0
+	for {
+		if subCursor == sub.Len() {
+			return true
+		}
+
+		if superCursor == super.Len() {
+			return false
+		}
+
+		subElem := sub.Elem(subCursor)
+		superElem := sub.Elem(superCursor + subCursor)
+		if superElem == nil {
+			return false
+		}
+
+		if subElem == nil {
+			subCursor = 0
+			superCursor++
+		}
+
+		if superElem.Value.Compare(subElem.Value) == 0 {
+			subCursor++
+		} else {
+			superCursor++
+			subCursor = 0
+		}
+	}
+
+	return false
+}
+
+func builtinObjectSubset(bctx BuiltinContext, operands []*ast.Term, iter func(*ast.Term) error) error {
+	superTerm := operands[0]
+	subTerm := operands[1]
+
+	if ok, superObj, subObj := bothObjects(superTerm, subTerm); ok {
+		// Both operands are objects.
+		return iter(ast.BooleanTerm(objectSubset(superObj, subObj)))
+	}
+
+	if ok, superSet, subSet := bothSets(superTerm, subTerm); ok {
+		// Both operands are sets.
+		return iter(ast.BooleanTerm(setSubset(superSet, subSet)))
+	}
+
+	if ok, superArray, subArray := bothArrays(superTerm, subTerm); ok {
+		// Both operands are sets.
+		return iter(ast.BooleanTerm(arraySubset(superArray, subArray)))
+	}
+
+	return fmt.Errorf("both arguments object.subset must be of the same type")
```

Comment:
Perhaps we can return an ErrOperand here? https://github.com/charlesdaniels/opa/blob/fd4379b68a40554de9d9fb7ca8bd7dfbf648ee3f/topdown/builtins/builtins.go#L31-L34

Reply:
I like the idea of that, but there isn't necessarily a good way to determine which position is the "wrong" one. Like if I give an array and a set, which one is "wrong" and which one is "right"? 

I guess I can just cast a string to the `ErrOperand` type, so that way code that's checking for `ErrOperand` will work...